<a href="https://colab.research.google.com/github/Aditya0721/Recommender/blob/master/Recommender.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **# The main idea behind the generation of the Recommendation system is to recommend posts using both Collaborative and Content Based Filtering Methods. For Collaborative based filtering we will find out all the categories and find the cosine similarity using user-user collaborative filtering.**
```



# **DATA PREPROCESSING**

## UPLOADING DATASETS AND STORING THEM IN DATAFRAMES 

In [0]:
import pandas as pd
from google.colab import files
import io

In [2]:
# upload posts.csv, users.csva and views .csv files to proceed
uploaded = files.upload()

Saving posts.csv to posts.csv
Saving users.csv to users.csv
Saving views.csv to views.csv


In [3]:
# creating a user data_frame
users_df = pd.read_csv('users.csv')
users_df.head()

_id                  name gender      academics
0  5d60098a653a331687083238  Nivesh Singh Chauhan   male  undergraduate
1  5d610ae1653a331687083239         Gaurav Sharma   male       graduate
2  5d618359fc5fcf3bdd9a0910         Akshay Mishra   male  undergraduate
3  5d6d2bb87fa40e1417a49315        Saksham Mathur   male  undergraduate
4  5d7c994d5720533e15c3b1e9         Varun Chowhan   male  undergraduate

In [4]:
# creating posts data frame
posts_df = pd.read_csv('posts.csv')
posts_df.head()

_id  ...  post_type
0  5d62abaa65218653a132c956  ...       blog
1  5d6d39567fa40e1417a4931c  ...       blog
2  5d7d23315720533e15c3b1ee  ...       blog
3  5d7d405e5720533e15c3b1f3  ...    artwork
4  5d80dfbc6c53455f896e600e  ...       blog

[5 rows x 4 columns]

In [5]:
# creating a views data frame
views_df = pd.read_csv('views.csv')
views_df.head()

user_id                   post_id                 timestamp
0  5df49b32cc709107827fb3c7  5ec821ddec493f4a2655889e  2020-06-01T10:46:45.131Z
1  5ed3748576027d35905ccaab  5ed4cbadbd514d602c1531a6  2020-06-01T09:39:20.021Z
2  5ed0defa76027d35905cc2de  5eac305f10426255a7aa9dd3  2020-06-01T08:12:42.682Z
3  5ed0defa76027d35905cc2de  5ed1ff0276027d35905cc60d  2020-06-01T08:10:23.880Z
4  5ed0defa76027d35905cc2de  5ed3820f76027d35905ccac8  2020-06-01T08:08:54.124Z

## DETAILS OF THE DATAFRAMES

In [6]:
users_df.describe()

_id              name gender      academics
count                        118               118    118            118
unique                       118               118      3              3
top     5e783375cfc8b713f5ac7c5b  Saumik Satapathy   male  undergraduate
freq                           1                 1     72             68

In [7]:
views_df.describe()

user_id  ...                 timestamp
count                       1449  ...                      1449
unique                       118  ...                      1449
top     5d60098a653a331687083238  ...  2020-05-18T16:40:18.228Z
freq                         230  ...                         1

[4 rows x 3 columns]

In [8]:
posts_df.describe()

_id             title     category  post_type
count                        493               493          465        493
unique                       493               477          231          4
top     5e895d87a3258347b42f25bb  PENCIL RENDERING  Photography    artwork
freq                           1                 3           81        241

## NULL VALUE IMPUTATION

In [9]:
users_df.isnull().any()


_id          False
name         False
gender       False
academics    False
dtype: bool

In [10]:
views_df.isnull().any()

user_id      False
post_id      False
timestamp    False
dtype: bool

In [11]:
posts_df.isnull().any()

_id           False
title         False
category       True
 post_type    False
dtype: bool

**we have got null values in the category column of the posts dataset**

**as its a categorical variable i have to fill the null values considering the title and post_type**

**counting total number of null values in posts_df**

In [12]:
posts_df['category'].isna().sum()

28

**finding title and post_type for all 28 null values**

In [13]:
# using pd.isna() function to find null values
posts_df[posts_df['category'].isna()]

_id  ...  post_type
24   5ddeb6e80eb5e25a8a07f065  ...    project
25   5de179d80eb5e25a8a07f079  ...    project
29   5dee9b5042a8854bf6eabaaf  ...    project
65   5e3ea110eab55d319938a7a7  ...    project
88   5e4c3873f5561b1994c8e3d9  ...    project
93   5e4da502f5561b1994c8e42c  ...    project
94   5e4da7c8f5561b1994c8e439  ...    project
95   5e4dab15f5561b1994c8e446  ...    project
96   5e4ed85af5561b1994c8e470  ...    project
97   5e4ed8ccf5561b1994c8e47d  ...    project
98   5e4edd1bf5561b1994c8e48a  ...    project
99   5e4edd87f5561b1994c8e497  ...    project
227  5e8c2d01a3258347b42f2627  ...    project
228  5e8c993da3258347b42f264d  ...    project
229  5e8cb88ea3258347b42f267e  ...    project
230  5e8cc876a3258347b42f26a3  ...    project
231  5e8ccfb5a3258347b42f26bc  ...    project
352  5ea5aacd10426255a7aa9b71  ...    project
353  5ea5ce9310426255a7aa9b8d  ...    project
374  5eaf3d4110426255a7aa9f4c  ...       blog
394  5eb1df6d10426255a7aaa029  ...    project
442  5eca8fceeaff6b0c3a58a3c0  ...    project
463  5ed0e31a76027d35905cc302  ...    project
464  5ed0e65976027d35905cc356  ...    project
465  5ed0f0c576027d35905cc3c7  ...    project
468  5ed175f176027d35905cc554  ...    project
469  5ed175f476027d35905cc562  ...    project
490  5ed3ef4cbd514d602c1530f2  ...    project

[28 rows x 4 columns]

**As we can see most of the null values are for the post_types project we will create a new category called project and assign it to the null values and when ever a user leaves the category field empty we will assign the category field same as the post_type as default**

In [0]:
posts_df[posts_df['category'].isna()]='project'

**As we have imputed the null values lets check again**

In [15]:
posts_df.isna().any()

_id           False
title         False
category      False
 post_type    False
dtype: bool

**Now lets see how many unique categories are present in the dataset**

In [16]:
posts_df['category'].unique()

array(['Plant Biotechnology',
       'Artificial Intelligence|Machine Learning|Information Technology',
       'Operating Systems', 'Drawings', 'Competition Laws', 'Eco System',
       'Economic Policies', 'Graphic|Graphic Design', 'Painting',
       'Pen and ink', 'Computer Technology|Information Technology',
       'Drawings|Painting',
       'Graphic Design|Visual Arts|Illustration|Graphic',
       'Drawings|Calligraphy', 'Photography', 'Empowerment', 'project',
       'Video editing', 'Inorganic Chemistry', 'Programming languages',
       'Conceptual|Graphic Design', 'HR Management',
       'Human Resources|HR Management',
       'Mass Media|International Relations', 'Sculptures|Artistic design',
       'Fashion Design|Ceramics|Artistic design', 'Craft|Artistic design',
       'Fashion Design|Visual Arts|Conceptual|Artistic design',
       'Photography|Fashion Design|Visual Arts|Graphic Design|Artistic design',
       'Fashion Design|Visual Arts|Graphic Design|Artistic design|Graph

## PREPARING MORE DETAILED DATASET

**As we can see some of the categories are combination of multiple categoris we will add more rows to the posts_df by splitting the categories and making one category for each row**

In [17]:
# creating a dictionary whcih will store actual categories as key and splitted 
# categories as its value
categories = {}
# filling the dictionary
for i in posts_df['category']:
  categories.update({i:[]})
  for j in list(i.split("|")):
      categories[i].append(j) 
print(categories)

{'Plant Biotechnology': ['Plant Biotechnology'], 'Artificial Intelligence|Machine Learning|Information Technology': ['Artificial Intelligence', 'Machine Learning', 'Information Technology'], 'Operating Systems': ['Operating Systems'], 'Drawings': ['Drawings'], 'Competition Laws': ['Competition Laws'], 'Eco System': ['Eco System'], 'Economic Policies': ['Economic Policies'], 'Graphic|Graphic Design': ['Graphic', 'Graphic Design'], 'Painting': ['Painting'], 'Pen and ink': ['Pen and ink'], 'Computer Technology|Information Technology': ['Computer Technology', 'Information Technology'], 'Drawings|Painting': ['Drawings', 'Painting'], 'Graphic Design|Visual Arts|Illustration|Graphic': ['Graphic Design', 'Visual Arts', 'Illustration', 'Graphic'], 'Drawings|Calligraphy': ['Drawings', 'Calligraphy'], 'Photography': ['Photography'], 'Empowerment': ['Empowerment'], 'project': ['project'], 'Video editing': ['Video editing'], 'Inorganic Chemistry': ['Inorganic Chemistry'], 'Programming languages': [

**Now lets update posts_df**

In [0]:
# we will prepare a list which will store all the new rows and
# column values for now 
updated_data =  []
for i in categories:
  # create a dummy dataframe which will store row values for the current 
  # category
  dummy = posts_df[posts_df['category']==i]
  id = dummy['_id'].values[0]
  title = dummy['title'].values[0]
  post_type = dummy[' post_type'].values[0]
  for j in categories[i]:
      # now we will create a dictionary which will store the row values of the 
      # main category and will be assigned to all of its splitted category 
      dict1 = {}
      dict1.update({'_id':id})
      dict1.update({'title':title})
      dict1.update({'category':j})
      dict1.update({' post_type':post_type})
      updated_data.append(dict1)

In [0]:
posts_df_updated = pd.DataFrame(updated_data)


**now lets check updated posts_df**

In [20]:
posts_df_updated.loc[:, ['_id', 'category', ' post_type']].head()

_id                 category  post_type
0  5d62abaa65218653a132c956      Plant Biotechnology       blog
1  5d6d39567fa40e1417a4931c  Artificial Intelligence       blog
2  5d6d39567fa40e1417a4931c         Machine Learning       blog
3  5d6d39567fa40e1417a4931c   Information Technology       blog
4  5d7d23315720533e15c3b1ee        Operating Systems       blog

In [21]:
posts_df.loc[:, ['_id', 'category', ' post_type']].head()

_id  ...  post_type
0  5d62abaa65218653a132c956  ...       blog
1  5d6d39567fa40e1417a4931c  ...       blog
2  5d7d23315720533e15c3b1ee  ...       blog
3  5d7d405e5720533e15c3b1f3  ...    artwork
4  5d80dfbc6c53455f896e600e  ...       blog

[5 rows x 3 columns]

In [22]:
posts_df_updated.describe()

_id       title             category  post_type
count                        592         592                  592        592
unique                       232         228                  235          4
top     5ecb72c0eaff6b0c3a58a48e  MOSAIC ART  Computer Technology       blog
freq                           7          10                   24        374

**As we can see we have successfully created our posts dataframe now we will merge the dataframes**

##MERGING OF THE DATAFRAMES 

**We will merge views_df and posts_df_updated dataframes as they conatin all the columns and data that we will need for our recommendation system**

In [23]:
views_df.columns

Index(['user_id', 'post_id', 'timestamp'], dtype='object')

In [24]:
posts_df_updated.columns

Index(['_id', 'title', 'category', ' post_type'], dtype='object')

**As we know for merging we need a common column and post_id is the column in the both dataframes but we have to rename one of them to perform the merge operation**
**We will change rename _id to post id in the posts_df_updated column** 

In [0]:
# using pd.rename function to rename _id to post_id
posts_df_updated.rename(columns={'_id':'post_id'}, inplace=True)

In [26]:
posts_df_updated.columns

Index(['post_id', 'title', 'category', ' post_type'], dtype='object')

In [0]:
# using pd.merge function to merge two dataframes
main_df = pd.merge(views_df, posts_df_updated)

In [28]:
main_df.columns

Index(['user_id', 'post_id', 'timestamp', 'title', 'category', ' post_type'], dtype='object')

In [29]:
main_df.head(10)

user_id  ...  post_type
0  5df49b32cc709107827fb3c7  ...    artwork
1  5df49b32cc709107827fb3c7  ...    artwork
2  5df49b32cc709107827fb3c7  ...    artwork
3  5df49b32cc709107827fb3c7  ...    artwork
4  5df49b32cc709107827fb3c7  ...    artwork
5  5ec3ba5374f7660d73aa1201  ...    artwork
6  5ec3ba5374f7660d73aa1201  ...    artwork
7  5ec3ba5374f7660d73aa1201  ...    artwork
8  5ec3ba5374f7660d73aa1201  ...    artwork
9  5ec3ba5374f7660d73aa1201  ...    artwork

[10 rows x 6 columns]

In [30]:
main_df.tail(20)

user_id  ...  post_type
1730  5d610ae1653a331687083239  ...       blog
1731  5d610ae1653a331687083239  ...       blog
1732  5d610ae1653a331687083239  ...       blog
1733  5d610ae1653a331687083239  ...       blog
1734  5d610ae1653a331687083239  ...       blog
1735  5e5855ced701ab08af792b51  ...       blog
1736  5e5855ced701ab08af792b51  ...       blog
1737  5e5855ced701ab08af792b51  ...       blog
1738  5e5855ced701ab08af792b51  ...       blog
1739  5e5855ced701ab08af792b51  ...       blog
1740  5d60098a653a331687083238  ...    artwork
1741  5d60098a653a331687083238  ...    artwork
1742  5d618359fc5fcf3bdd9a0910  ...       blog
1743  5d618359fc5fcf3bdd9a0910  ...       blog
1744  5e7def80a3258347b42f2124  ...       blog
1745  5d60098a653a331687083238  ...       blog
1746  5d60098a653a331687083238  ...       blog
1747  5d60098a653a331687083238  ...    artwork
1748  5d60098a653a331687083238  ...    artwork
1749  5d60098a653a331687083238  ...    artwork

[20 rows x 6 columns]

In [31]:
main_df.describe()

user_id  ...  post_type
count                       1750  ...       1750
unique                        88  ...          3
top     5d60098a653a331687083238  ...       blog
freq                         261  ...        846

[4 rows x 6 columns]

**Now the merging is done and we have our actuall dataframe to work on now lets move to the nect step which is filtering using Collaborative Filtering**

# **Collaborative Filtering**

**So we are using user-user collaberative filtering. And we are using cosine similarity to determine the similarity between users. For that we need to build a matrix. categories will be the columns and user_ids will be the rows. And we are using K near neighbour model to determine the similar users.**

## CONSTRUCTING THE MATRIX

**We will create a sparse matrix of mXn. All the unique categories will be the column values and user_ids will be the row value. And the cell values will be the number times a user has viwed a category. The higher the cell value the higher the users interest for that particular category**

**Lets find all the unique user_ids and categories**

In [32]:
users = list(main_df['user_id'].unique())
len(users)

88

In [33]:
categories = list(main_df['category'].unique())
len(categories)

234

**Our matrix size will be of size 88X234**

In [0]:
# craeting the matrix
user_mat = [[] for i in range(len(users))]
for i in range(len(users)):
    for j in range(len(categories)):
        # finding how many rows present with current user_id and current
        # category
        value = len(main_df[
                    (main_df['user_id']==users[i])&
                    (main_df['category']==categories[j])
                    ])
        user_mat[i].append(value)

In [35]:
for i in user_mat[0]:
    print(i, end=" ")

4 3 3 3 2 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 2 0 0 0 0 1 2 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 

**Now we will convert the list to a actual sparse matrix using scipy.sparse library**

In [0]:
# importing csr_matrix from scipy.sparse
from scipy.sparse import csr_matrix # for creating a sparse matrix
user_mat = csr_matrix(user_mat)

In [37]:
user_mat

<88x234 sparse matrix of type '<class 'numpy.longlong'>'
	with 1034 stored elements in Compressed Sparse Row format>

**As we have made our matrix lets train the model.**

## TRAINING THE MODEL

**So we are using K nearest neighbour model. And the similarity constraint we are using is cosine similarity.**

**Importing the model**

In [0]:
# importing NearestNeighbors from sklearn.neighbors
from sklearn.neighbors import NearestNeighbors

In [0]:
# setting the parameters
model_knn = NearestNeighbors(
                metric='cosine', algorithm='brute', n_neighbors=15)

In [40]:
# fitting the model
model_knn.fit(user_mat)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
                 metric_params=None, n_jobs=None, n_neighbors=15, p=2,
                 radius=1.0)

## CREATING A FUNCTION FOR RECOMMENDATION

**We have trained our model. Now lets make a function which will take users_id as input and recommend categories that the user would like**

In [0]:
# creating a function which will recommend us posts with argument as
# user_id and other default arguments
def recommender(user_id, data=user_mat, model=model_knn):
    model.fit(data)
    index = users.index(user_id)
    current_user = main_df[main_df['user_id']==user_id]
    distances, indices = model.kneighbors(data[index], 15)
    recomendation = []
    for i in indices[0]:
        user = main_df[main_df['user_id']==users[i]]
        for i in user['category'].unique():
            if i not in current_user['category'].unique():
                recomendation.append(i)
    print(recomendation)
    print(indices)

**What we did above there is first we found out 15 similar users for our current user and we recomended those posts to the user which he/she hasnot viewed yet but similar users have.Now lets test our function for a user**

**Lets see what the system recomendation for the first user. But before that lets see the users categories that he has already viewed**

In [42]:
main_df[main_df['user_id']==users[0]]['category'].unique()

array(['Visual Arts', 'Graphic Design', 'Artistic design', 'Graphic',
       'Illustration', 'Photography', 'Architecture', 'Logo Design',
       'Art; Science', 'Science;Technology', 'Fashion Desigining',
       'Pattern & Culture', 'Fashion Textile', 'Fashion Trends',
       'Fashion Illustration', 'Pattern Cutting', 'Fashion Communication',
       'Art', 'Calligraphy'], dtype=object)

**These are the categories the user has viewed now lets see the recomendations for our user**

In [43]:
recommender(users[0])

['Fashion Portfoilio', 'Fashion Design', 'Watercolours', 'Drawings', 'Sketch Video', 'Political Science', 'Colonialism In India', 'Legal Studies', 'Labor Law', 'Banking', 'Banking Companies', 'Professionalism', 'Fashion Portfoilio', 'Painting', 'Craft', 'Technology', 'Drawings', 'Business', 'Learning', 'Literature', 'Stories', 'Fictions', 'Movements in Literature', 'Fashion Portfoilio', 'Business Strategies', 'Business Enviorment', 'New Venture Planning', 'Foreign Business', 'Business Organisation', 'Fashion Design', 'Watercolours', 'Conceptual', 'Pen and ink', 'Mixed Media', '2D Composition', 'Communication', 'Basics of Communiaction', 'Sketch Video', 'Professionalism', 'Biotechnology', 'Molecular Biology', 'Fine Arts', 'Painting', 'Management', 'Business Management', 'Technology Mangement', 'Managerial Activity', 'Creative And Lateral Management', 'Economics', 'Economic Policies', 'Break even Point', 'Foreign Economy', 'Computer Technology', 'Operating Systems', 'Fashion Manufacturin

**NOTE: We are suggesting what are the new categories the user would like we can also modify the recommend function to suggest Posts also.**

**Looks like the system is suggesting posts of users interest and also some new categories for the user.**
**The List below suggests the similar users to the current user.**

**So This ends our Collaborative Filtering. Lets move on to Content based filtering**

# CONTENT BASED FILTERING

**Now for content based filtering we will prepare item profiles for each category and user profile for each user.**

## PREPARING ITEM-PROFILES



**We have got 234 unique categories. Each category will have its own item-profile. These item-profiles are vectors of length equal to number of posts we have and it will contain binary values depending upon whether the category is present in a post or not.**

In [48]:
main_df['post_id'].describe()

count                         1750
unique                         231
top       5ec7abfdec493f4a26558860
freq                            63
Name: post_id, dtype: object

**We will have 234 vectors, each of length 231 containing 0s and 1s**

In [0]:
imoport numpy as np

# Lets create a list which will contain all the posts
posts = list(main_df['post_id'].unique())

# Create a dictionary which will store each category as key and their item-profile vector as value 
item_profiles = {}
# filling item_profiles
for i in categories:
    item_profiles.update({i:[]})
    for j in posts:
        item_profiles[i].append(1) if i in list(main_df[main_df['post_id']==j]['category'].unique()) else item_profiles[i].append(0)

In [92]:
# converting lists to vectors or arrays
for i in item_profiles:
    item_profiles[i] = np.array(item_profiles[i])
for i in item_profiles:
    print(i, item_profiles[i])

Visual Arts [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 1 0 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 0]
Graphic Design [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 0]
Artistic design [1 0 0 0 0 0 0 0 

## PREPARING THE USER-PROFILE

**We will prepare the user profile from the item profiles that we have already prepared. Lets say a user has viewed posts which have n different categories and let I1,I2....In be their item-profiles. User profile for a user will be equal to the weghited average of the profiles. For the weight calculation we will divide the number of times a category has appeared in all the posts the user have viewed by the total number of pots the user has viewed.**

In [0]:
# we will create a dictionary user_id as key and user_profile as value
user_profiles = {}
# Filling the user_profiles
for i in users:
    user_profiles.update({i:[]})
    # finding the current user
    current_user = main_df[main_df['user_id']==i]
    # listing all the categories the user has viewed
    current_user_categories = list(current_user['category'].unique())
    # Listing all the posts the user has viewed
    current_user_post = list(current_user['post_id'].unique())

    # Now we will find the weights for each category in the current_user_categories
    # create a dictionary category as key and their weight as value
    category_weight = {}
    # create a dummy vector which will store the final vector for the user profile of length equal to posts
    result_vector = np.array([0 for i in range(len(posts))])
    for j in current_user_categories:
        category_weight.update({j:0})
        # Now count how many times j has appeared
        for k in list(current_user['category']):
            if j==k:
                category_weight[j]+=1
        # Now divide it with the length of the current_user_post
        category_weight[j] = category_weight[j]/len(current_user_post)
        # Now we have calculated our weights, Now we will calculate user-profile
        result_vector = result_vector+ (category_weight[j]*item_profiles[j])
    user_profiles[i] = result_vector/len(current_user_post)

In [105]:
for i in user_profiles:
    if i in users[0:5]:
        print(i, user_profiles[i])
        print()

5df49b32cc709107827fb3c7 [0.12396694 0.         0.         0.         0.00826446 0.00826446
 0.         0.00826446 0.00826446 0.         0.         0.
 0.         0.         0.         0.         0.         0.04132231
 0.         0.15702479 0.         0.00826446 0.09090909 0.02479339
 0.03305785 0.         0.         0.02479339 0.01652893 0.01652893
 0.01652893 0.01652893 0.         0.05785124 0.07438017 0.
 0.00826446 0.         0.         0.         0.         0.
 0.08264463 0.01652893 0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.04132231 0.04132231 0.         0.00826446 0.
 0.         0.         0.02479339 0.00826446 0.         0.07438017
 0.         0.         0.03305785 0.         0.         0.
 0.         0.09917355 0.         0.09917355 0.         0.
 0.03305785 0.03305785 0.         0.         0.024

**So we have successfully Calculated the user_profile and item_profile.**

## CREATING A FUNCTION FOR RECOMMENDATION

**In this Section we will find the cosine similarity between a user and the categoroies, the higher the cosine value the higher will be the similarity**

In [0]:
# to find the cosine similarity we need to import cosine_similarity function
from sklearn.metrics.pairwise import cosine_similarity

# we will create a function which will take user_id as argument and will provide recomendation
# all other arguments will be set to some default values. 
def recommender1(user_id, user_profiles = user_profiles, item_profiles=item_profiles):
    # calculate the cosine similiraity between the item-profile of all categories and the users user-profile
    # create a dictionary category as key and cosine value as the value
    similarity = {}
    for i in item_profiles:
        similarity.update({i:cosine_similarity(user_profiles[user_id].reshape(1, -1), item_profiles[i].reshape(1, -1))})
    # as we have found the similarity now we will sort it and show the user the posts of the top categories and which the user
    # hasnot viewed yet
    sorted_similarity = sorted(similarity.items(), key=lambda x: x[1], reverse=True)
    # now we will filter those posts that the user hasnot viewed
    user_posts = list(main_df[main_df['user_id']==user_id]['post_id'].unique())
    # create recomendation list
    recommendations = []
    # displaying users viewed posts and categories
    print("posts user has viewed:{}".format(user_posts))
    print("users viewed categories:{}".format(list(main_df[main_df['user_id']==user_id]['category'].unique())))
    for i in sorted_similarity:
        category_posts = list(main_df[main_df['category']==i[0]]['post_id'].unique())
        for j in category_posts:
            if j not in user_posts:
                recommendations.append([i[0], j])
        # we will recommend top 20 posts to the user
        if len(recommendations)==20:
            break
    for i in recommendations:
        print(i)

In [154]:
recommender1(users[30])

posts user has viewed:['5ea7cd9610426255a7aa9bd2']
users viewed categories:['Business']
['Business', '5ea80a7e10426255a7aa9be2']
['Business', '5eadc2f710426255a7aa9ee4']
['Business', '5e4d359cf5561b1994c8e424']
['Business', '5e6eff8aed32a005135d6bb8']
['Business', '5e830dc8a3258347b42f23fb']
['Business', '5e978c7ca3258347b42f2b09']
['Business', '5e9028cea3258347b42f2736']
['Business', '5e8e13dca3258347b42f26f2']
['Business Strategies', '5ea80a7e10426255a7aa9be2']
['Business Strategies', '5e978c7ca3258347b42f2b09']
['Business Strategies', '5e8e13dca3258347b42f26f2']
['Venture Capitalist', '5eadc2f710426255a7aa9ee4']
['Venture Capitalist', '5e978c7ca3258347b42f2b09']
['Venture Capitalist', '5e8e13dca3258347b42f26f2']
['Business Enviorment', '5ea80a7e10426255a7aa9be2']
['Business Enviorment', '5eadc2f710426255a7aa9ee4']
['New Venture Planning', '5ea80a7e10426255a7aa9be2']
['New Venture Planning', '5eadc2f710426255a7aa9ee4']
['Business Organisation', '5ea80a7e10426255a7aa9be2']
['Business 

**Looks like we are getting pretty satisfying results**

**SO THIS CONCLUDES MY RECOMMENDATION SYSTEM USING COLLABORATIVE AND 
CONTENT BASED FILTERING**